app/main/services/query_builder.py

```diff
@@ -73,34 +166,26 @@ def match_all_clause():
 
 
 def filter_clause(query_args):
-    return {
-        "bool": {
-            "must": build_term_filters(query_args)
-        }
+    and_filters = []
+    or_filters = []
+
+    query_filters = [QueryFilter(field, values)
+                     for field, values in query_args.lists()
+                     if field.startswith("filter")]
+    filters = {
+        "bool": {}
     }
 
+    for each_filter in query_filters:
+        if each_filter.is_and_filter():
+            and_filters += each_filter.terms()
+        if each_filter.is_or_filter():
+            or_filters += each_filter.terms()
 
-def build_term_filters(query_args):
-    must = []
-    if "serviceTypes" in query_args:
-        for service_type in extract_service_types(query_args):
-            must.append({
-                "term": {
-                    "serviceTypesExact":
-                        strip_and_lowercase(service_type)
-                }
-            })
-    if "lot" in query_args:
-        must.append({
-            "term": {
-                "lot": query_args["lot"]
-            }
-        })
-    return must
+    if and_filters:
+        filters["bool"]["must"] = and_filters
 
+    if or_filters:
+        filters["bool"]["should"] = or_filters
```

Comment:
This is mostly right, but I don't think `minimum_should_match` is valid in that context (I think it'll be ignored).  The documentation of it isn't clear, but I think it's a parameter that's used for textual queries to determine how many tokens need to match.  This would do what you're wanting without it being specified, though.

There's an alternate approach that I'd tend to use: elasticsearch supports "or" and "and" filters (see http://www.elastic.co/guide/en/elasticsearch/reference/1.5/query-dsl-or-filter.html)  These just combine a list of filters with an OR or an AND operator, and are much easier to understand than what you get by using the bool filter, IMO.

In rummager, we use a helper to take a list of filters and combine them with an OR or AND operator (with special cases for lists of length 0 or 1): see https://github.com/alphagov/rummager/blob/master/lib/unified_search_builder.rb#L150.  This ends up producing code which you can read pretty much as a boolean expression.